src/core/or/circuitpadding_machines.c

```diff
@@ -0,0 +1,351 @@
+ * this machine, otherwise it's the service-side of this machine. */
+static void
+setup_burst_state_for_hiding_rend_circuits(circpad_state_t *burst_state,
+                                           bool is_client)
+{
+  /* Token removal strategy for BURST state */
+  burst_state->token_removal =
+      CIRCPAD_TOKEN_REMOVAL_CLOSEST;
+
+  /* Histogram for BURST state:
+   *
+   * We want a machine that will send a bunch of fake cells over a reasonable
+   * time frame, to obfuscate the fact that introduction circuits always send 4
+   * cells both ways */
+
+  /* Figure out the length of the BURST state so that it's randomized. See
+     definition of these constants for rationale. */
+  burst_state->length_dist.type = CIRCPAD_DIST_UNIFORM;
+  burst_state->length_dist.param1 = is_client ?
+    HS_MACHINE_PADDING_MINIMUM_CLIENT : HS_MACHINE_PADDING_MINIMUM_SERVICE;
+  burst_state->length_dist.param2 = is_client ?
+    HS_MACHINE_PADDING_MAXIMUM_CLIENT : HS_MACHINE_PADDING_MAXIMUM_SERVICE;
+
+  /* Histogram is: (30 msecs, 200 msecs, infinity).
+   * See analysis above.
+  */
+  burst_state->histogram_len = 2;
+  burst_state->histogram_edges[0] = 30000;
+  burst_state->histogram_edges[1] = 200000;
+
+  /* Maximum number of tokens for the BURST state. We control the BURST state
+   * duration based on the length_dist so the token count does not really
+   * matter, it just needs to be bigger than all possible lengths. */
+  burst_state->histogram[0] = 1000;
+  burst_state->histogram_total_tokens = 1000;
+}
+
+/** Create a client-side padding machine that aims to hide IP circuits. In
+ *  particular, it keeps intro circuits alive until a bunch of fake traffic has
+ *  been pushed through.
+ *
+ *  This aims to hide the fingerprint of client-side introduction circuits
+ *  which is the fact that they are short-lived and always send 4 cells to the
+ *  intro point (3 EXTEND and 1 INTRODUCE1) (section 5.1).
+ */
+void
+circpad_machine_client_hide_intro_circuits(smartlist_t *machines_sl)
+{
+  circpad_machine_spec_t *client_machine
+      = tor_malloc_zero(sizeof(circpad_machine_spec_t));
+
+  client_machine->name = "client_ip_circ";
+
+  client_machine->conditions.state_mask =
+    CIRCPAD_CIRC_BUILDING|CIRCPAD_CIRC_OPENED;
+  client_machine->target_hopnum = 2;
+
+  /* This is a client machine */
+  client_machine->is_origin_side = 1;
+
+  /* For this machine we only pad in client-side introduction circuits even
+   * after introduction has finished */
+  client_machine->conditions.purpose_mask =
+    circpad_circ_purpose_to_mask(CIRCUIT_PURPOSE_C_INTRODUCING)|
+    circpad_circ_purpose_to_mask(CIRCUIT_PURPOSE_C_INTRODUCE_ACK_WAIT)|
+    circpad_circ_purpose_to_mask(CIRCUIT_PURPOSE_C_INTRODUCE_ACKED);
+
+  /* XXX Keep the circuit alive even after the introduction has been finished,
+   * otherwise the short-term lifetime of the circuit will blow our cover */
+  // client_machine->manage_circ_lifetime = 1;
```

Comment:
Note that we also want to set client_machine->should_negotiate_end = 1 on the client side here, since the relay side will not be managing lifetime (as you say below).

Reply:
Done in 7dc7c02f13.